doc/click-xform.1
.\" -*- mode: nroff -*-
.ds V 1.0.3
.ds E " \-\- 
.if t .ds E \(em
.de Sp
.if n .sp
.if t .sp 0.4
..
.de Es
.Sp
.RS 5
.nf
..
.de Ee
.fi
.RE
.PP
..
.de Rs
.RS
.Sp
..
.de Re
.Sp
.RE
..
.de M
.BR "\\$1" "(\\$2)\\$3"
..
.de RM
.RB "\\$1" "\\$2" "(\\$3)\\$4"
..
.TH CLICK-XFORM 1 "31/Mar/2000" "Version \*V"
.SH NAME
click-xform \- pattern-based Click configuration optimizer
'
.SH SYNOPSIS
.B click-xform
.RI \%[ options ]
.RI \%[ param = value " ...]"
.RI \%[ router\-file " [" pattern\-files ...]]
'
.SH DESCRIPTION
The
.B click-xform
tool is a Click router configuration optimizer. It reads files giving Click
configuration patterns and their replacements, and a router configuration
file. Then it replaces pattern matches in the configuration until none
remain and writes the new configuration to the standard output.
.PP
Pattern files use the Click language (see
.M click 5 ).
They contain pairs of `elementclass' definitions, where each pair consists
of one pattern text and one replacement text. The replacement for a pattern
named
.RI ` X '
must be named
.RI ` X _Replacement'.
This example pattern file replaces all Queues with Queue-Shaper
combinations:
.Rs
.nf
elementclass QueuePattern {
  input -> Queue -> output;
}
elementclass QueuePattern_Replacement {
  input -> Queue -> Shaper(200) -> output;
}
.fi
.Re
.PP
A pattern text can contain an arbitrary number of elements, and an
arbitrary number of input and output ports. Basically, a pattern matches a
fragment of a router configuration when they have the same number of
elements; their element classes match; all of the pattern's connections are
duplicated in the fragment; and any connections into or out of the fragment
correspond to connections to or from the pattern's input and output ports.
The formal definition of matching is given below.
.PP
Any configuration strings in the pattern must match the configuration
strings in the configuration subset. The pattern can use variables, which
look like `$[letters, numbers and underscores]', to match a set of
configuration strings; except for variables, the strings must match
verbatim. A variable matches a single configuration argument. The same
variable can be used multiple times in a pattern; if so, then it must match
the same text on each occurrence. If a variable is used in the
replacement's configuration strings, then when a replacement is made, the
text that matched in the pattern will be inserted instead. For example,
applying this pattern
.Rs
.nf
elementclass Meters {
  input -> Meter($a)
        -> Shaper($b) -> output;
}
elementclass Meters_Replacement {
  input -> Meter($b)
        -> SlowShaper($a) -> output;
}
.fi
.Re
to this configuration
.Rs
.nf
\&... -> Meter(1000) -> Shaper(2000) -> ...
.fi
.Re
will create this result:
.Rs
.nf
\&... -> Meter(2000) -> SlowShaper(1000) -> ...
.fi
.Re
.PP
The optimizer will not apply the same pattern to a configuration subset
twice in succession. Specifically, every replacement element is marked with
the pattern that generated it; a pattern will not match a configuration
fragment if every element in that fragment came from that pattern. Thus, a
pattern like the QueuePattern above won't cause an infinite loop. You can
still cause an infinite loop, if you'd like, by having two patterns that
match the same text:
.Rs
.nf
elementclass Evil1 {
  input -> Queue -> output;
}
elementclass Evil1_Replacement {
  input -> Queue -> output;
}
elementclass Evil2 {
  input -> Queue -> output;
}
elementclass Evil2_Replacement {
  input -> Queue -> output;
}
.fi
.Re
This collection of patterns will make the optimizer run forever on any
configuration that has a Queue.
.PP
The
.B click-xform
transformation can be reversed with the
.B \-\-reverse
option.
'
.SH "OPTIONS"
'
If any filename argument is a single dash "-",
.B click-xform
will use the standard input or output instead, as appropriate.
'
.TP 5
.BI \-p " file"
.PD 0
.TP
.BI \-\-patterns " file"
Read patterns and replacements from
.IR file .
You can give any number of these options.
'
.Sp
.TP
.BI \-f " file"
.TP
.BI \-\-file " file"
Read the router configuration to transform from
.IR file .
The default is the standard input.
'
.Sp
.TP
.BI \-e " expr"
.TP
.BI \-\-expression " expr"
Use
.IR expr ,
a string in the Click language, as the router configuration to transform.
'
.Sp
.TP
.BI \-o " file"
.TP
.BI \-\-output " file"
Write the output router configuration to
.IR file .
The default is the standard output.
'
.Sp
.TP
.BR \-r ", " \-\-reverse
Apply the patterns in reverse. That is, replace occurrences of the
replacement texts with the corresponding pattern texts.
'
.Sp
.TP 5
.BI \-\-help
Print usage information and exit.
'
.Sp
.TP
.BI \-\-version
Print the version number and some quickie warranty information and exit.
'
.PD
'
.SH "FORMAL DEFINITION OF MATCHING"
'
A pattern
.I P
matches a subset
.I S
of the configuration's elements if the following conditions hold:
.TP 4
\(bu
There is a one-to-one mapping 
.I map
from
.I P
to
.I S
that respects element classes (that is, if an element
.IR p " in " P
has class
.IR K ,
then
.RI map( p ") also has class " K ).
.TP 4
\(bu
The configuration strings match, possibly by using a consistent variable
assignment.
.TP 4
\(bu
For every connection
.RI ` p1 " [" x "] -> [" y "] " p2 '
in the pattern
.IR P ,
there exists a connection
.RI `map( p1 ") [" x "] -> [" y "] map(" p2 )'
in the configuration subset
.IR S .
.TP 4
\(bu
For every connection
.RI ` c1 " [" x "] -> [" y "] " c2 '
in the configuration, one of four conditions hold:
.RS
.TP 3
\(bu
The connection is wholly outside the subset
.IR S .
(That is,
.IR c1 " is not in " S
and
.IR c2 " is not in " S .)
.TP 3
\(bu
The connection is inside the subset, and corresponds to a connection in the
pattern. (That is,
.IR c1 " is in " S ,
.IR c2 " is in " S ,
and
.I P
has a connection
.RI "`map-1(" c1 ") [" x "] -> [" y "] map-1(" c2 ")'.)"
.TP 3
\(bu
The connection goes into the subset, and corresponds to an input port in
the pattern. (That is,
.IR c1 " is not in " S
but
.IR c2 " is in " S ,
and there exists an input port number
.I i
so that
.I P
has a connection
.RI "`input [" i "] -> [" y "] map-1(" c2 ")',"
and for every connection in the pattern
.RI "`input [" i "] -> [" z "] " q ',
there is a connection in the configuration
.RI ` c1 " [" x "] -> [" z "] map(" q ")'.)"
.TP 3
\(bu
The connection goes out of the subset, and corresponds to an output port in
the pattern. (That is,
.IR c1 " is in " S
but
.IR c2 " is not in " S ,
and there exists an output port number
.I o
so that
.I P
has a connection
.RI "`map-1(" c1 ") [" x "] -> [" o "] output',"
and for every connection in the pattern
.RI ` q " [" z "] -> [" o "] output',"
there is a connection in the configuration
.RI "`map(" q ") [" z "] -> [" y "] " c2 "'.)"
.RE
'
.SH "SEE ALSO"
.M click 5
'
.SH AUTHOR
.na
Eddie Kohler, kohler@seas.harvard.edu
.br
http://www.pdos.lcs.mit.edu/click/
'


NEWS
Click NEWS

Version 2.0.1   24.Sep.2011

* ns-3 support, thanks to Lalith Suresh and Sascha Alexander Jopen.

* Important bug fixes for user-level ToDevice thanks to Jaeyong Yoo.

* Better handling for time: Timestamp knows about steady clocks (also known
  as monotonic clocks -- clocks that never go backwards) on most platforms,
  so Timers should go off at predictable intervals even when the system
  time is set backwards.

* IPRewriter and friends: TCP supports >2GB transfers and multiple sequence
  number transitions; UDP supports different timeouts for streaming
  connections and RPC-like connections.

* Configure with --enable-stats=2 for more precise cycle counting.

* Classifier and friends (IPFilter, etc.): Correct bugs and improve
  optimization.

* Bug fixes for AddressInfo, user-level packet pool (thanks to Seiichi
  Tetsukawa).

* Support packages on patchless Linux installs.

* Thanks to Cliff Frey, and to Felipe Chaulet, Sascha Alexander Jopen,
  Christian Kreibich, Meraki, Inc., Roberto Riggio, Jim Roewe, Robert
  Sombrutzki, Lalith Suresh, Dimitris Syrivelis, Seiichi Tetsukawa, Michael
  W.S. Yee, Jaeyong Yoo, and all other contributors.


Version 2.0   16.Jul.2011

* New elements: AdjustTimestamp, BlockThread, ClickyInfo, EtherVLANEncap,
  HandlerTask, ICMPIPEncap, NullTask, SetVLANAnno, StoreUDPTimeSeqRecord,
  StripDSRHeader, StripEtherVLANHeader, TCPFragmenter, UDPIP6Encap,
  UnstripDSRHeader.

* Improved elements: ARPQuerier, AddressInfo, BandwidthRatedSplitter,
  BandwidthRatedUnqueue, CheckLength, DRRSched, FromDevice linuxmodule,
  FromDevice userlevel, FromHost userlevel, FromIPSummaryDump,
  FromSimDevice, FromUserDevice, GetIPAddress, HostEtherFilter, IPFilter
  ("ether" comparisons, faster initialization), InfiniteSource, KernelTun,
  LinkUnqueue, RandomSource, RatedSource, RatedSplitter, RatedUnqueue,
  Script, SetIPAddress, SetTimestamp, TCPIPSend, TimeSortedSched, ToDevice
  userlevel, ToUserDevice, Truncate, Unqueue2, and others.

* Element and tool bug fixes: CheckTCPHeader, CheckUDPHeader, DSRRouteTable
  DelayUnqueue, FastUDPSourceIP6, FromHost, FromUserDevice, IP6Encap,
  IPAddrRewriter, IPFilter and other classifiers, KernelFilter,
  RadiotapDecap, RadixIPLookup, SetIP6DSCP, Socket, ThreadSafeQueue,
  TimeSortedSched, ToUserDevice, UDPIPEncap, WifiDecap, click-align,
  click-undead, and others.

* Performance improvements

** Tasks: A simplified reschedule implementation and streamlined driver
   loop is both faster and less error prone (now calling fast_reschedule()
   outside of Element::run_task() is safe).

** Tasks: If a task does no work, delay re-running it.  This simple change
   improves performance on many configurations, particularly those mixing
   file descriptor or timer access and Tasks.

** Timers: A 4-heap improves performance with very large numbers of timers.

** User-level: Device elements default to libpcap even on Linux, offering
   improved performance depending on how libpcap was compiled (e.g. packet
   mmap support).  A packet memory pool suggested by Luigi Rizzo offers
   much better performance for packet allocation and freeing.

** The round-robin scheduler (--disable-stride) works again.

** New configure option --enable-bound-port-transfer slightly optimizes
   push() and pull() calls (Cliff Frey).

* Language changes

** New "require(library FILE)" allows one Click script to include another
   (Christian Kreibich).

** Many-to-one, one-to-many, and many-to-many connections cleanly express
   some connection patterns.  For example, "c :: Classifier => a1, a2, a3"
   means the same as "c :: Classifier; c [0] -> a1; c [1] -> a2; c [2] ->
   a3".

** Element groups cleanly express short detours from a connection path.
   For example, this:

	c :: Classifier(00/01);
	x -> c -> y;
	c [1] -> Paint(1) -> y;

   can be expressed like this:

	x -> c :: Classifier(00/01) => (
            [0] -> [0]; [1] -> Paint(1) -> [0]
        ) -> y;

* Multithreading

** Many performance and stability improvements to multithreaded
   configurations, particularly at userlevel.

** Lazy Task rescheduling reduces opportunities for bugs.

** Each thread gains its own sets of Timers (and, at user level, selector
   file descriptors).  An element's Timers are initially scheduled on the
   same thread as its Tasks.  This should simplify reasoning about
   Task/Timer conflicts, and can help improve performance.

* Linuxmodule

** Patchless installs are on by default.  Click installs and works without
   a patch on many versions of Linux up to 2.6.38.

* Clicky changes

** Per-router styles: ClickyInfo elements let a script specify how it
   should be drawn.

** "Run" and "Stop" buttons can execute a configuration.

* Internals

** Args: This new class and its helpers support concise, easy-to-use,
   type-safe argument parsing and are strongly preferred to cp_va_kparse in
   new code.

** TokenBucket: This new template class supports fast, flexible,
   integer-based token bucket rate management.

** Timestamps default to nanosecond precision on many platforms.

** Element::simple_action() can be used for multi-port elements.

** On some platforms, set environment variable CLICK_BACKTRACE to 1 and get
   a useful stack backtrace if Click crashes (Cliff Frey).

** Other improvements: more flexible handlers, simple_action, type traits,
   faster multi-precision arithmetic, better IPv6 address unparsing.

* Acknowledgements

** Thanks to Cliff Frey for extensive discussion, multithreaded debugging,
   802.1Q elements, TokenBucket, TCPFragmenter, bound port transfer, and
   others.

** Thanks to Joonwoo Park for Linux kernel module patches.

** Thanks to Jimmy Kjällman and Pekka Nikander for FreeBSD kernel module
   support.

** Thanks to Johan Bergs, Jesse Brown, Roman Chertov, Beyers Cronje, Jens
   De Wit, Kevin Paul Herbert, Sascha Alexander Jopen, Christian Kreibich,
   Bobby Longpocket, Ruben Merz, Derrick Pallas, Roberto Riggio, Luigi

tools/click-xform/click-xform.cc
/*
 * click-xform.cc -- pattern-based optimizer for Click configurations
 * Eddie Kohler
 *
 * Copyright (c) 1999-2000 Massachusetts Institute of Technology
 * Copyright (c) 2007 Regents of the University of California
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, subject to the conditions
 * listed in the Click LICENSE file. These conditions include: you must
 * preserve this copyright notice, and you cannot mention the copyright
 * holders in advertising related to the Software without their permission.
 * The Software is provided WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED. This
 * notice is a summary of the Click LICENSE file; the license in that file is
 * legally binding.
 */

#include <click/config.h>

#include "routert.hh"
#include "lexert.hh"
#include <click/error.hh>
#include <click/confparse.hh>
#include <click/variableenv.hh>
#include <click/driver.hh>
#include <click/clp.h>
#include "toolutils.hh"
#include "adjacency.hh"
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>

bool match_config(const String &, const String &, HashTable<String, String> &);
// TODO: allow some special pports to be unconnected

class Matcher { public:

  Matcher(RouterT *, AdjacencyMatrix *, RouterT *, AdjacencyMatrix *, int, ErrorHandler *);
  ~Matcher();

  bool check_into(const PortT &, const PortT &);
  bool check_out_of(const PortT &, const PortT &);

  bool check_match();
  bool next_match();

  void replace_config(ElementT *) const;
  void replace(RouterT *, const String &, const LandmarkT &, ErrorHandler *);

 private:

  RouterT *_pat;
  AdjacencyMatrix *_pat_m;
  RouterT *_body;
  AdjacencyMatrix *_body_m;
  int _patid;

  ElementT *_pat_input;
  ElementT *_pat_output;

  Vector<ElementT *> _match;
  Vector<ElementT *> _back_match;
  HashTable<String, String> _defs;

  Vector<PortT> _to_pp_from;
  Vector<PortT> _to_pp_to;
  Vector<PortT> _from_pp_from;
  Vector<PortT> _from_pp_to;

};


Matcher::Matcher(RouterT *pat, AdjacencyMatrix *pat_m,
		 RouterT *body, AdjacencyMatrix *body_m,
		 int patid, ErrorHandler *errh)
  : _pat(pat), _pat_m(pat_m), _body(body), _body_m(body_m), _patid(patid),
    _pat_input(0), _pat_output(0)
{
  // check tunnel situation
  for (RouterT::iterator x = _pat->begin_elements(); x; x++) {
    if (!x->tunnel())
      /* nada */;
    else if (x->tunnel_connected())
      errh->lerror(x->landmark(), "pattern has active connection tunnels");
    else if (x->name() == "input" && !_pat_input)
      _pat_input = x.get();
    else if (x->name() == "output" && !_pat_output)
      _pat_output = x.get();
    else
      errh->lerror(x->landmark(), "connection tunnel with funny name '%s'", x->name_c_str());
  }
}

Matcher::~Matcher()
{
}

bool
Matcher::check_into(const PortT &houtside, const PortT &hinside)
{
  PortT phinside(_back_match[hinside.eindex()], hinside.port);
  PortT success;
  // now look for matches
  for (RouterT::conn_iterator it = _pat->find_connections_to(phinside);
       it != _pat->end_connections(); ++it)
    if (it->from_element() == _pat_input
	&& (success.dead() || it->from() < success)) {
      Vector<PortT> pfrom_phf, from_houtside;
      // check that it's valid: all connections from tunnels are covered
      // in body
      _pat->find_connections_from(it->from(), pfrom_phf);
      _body->find_connections_from(houtside, from_houtside);
      for (int j = 0; j < pfrom_phf.size(); j++) {
	PortT want(_match[pfrom_phf[j].eindex()], pfrom_phf[j].port);
	if (want.index_in(from_houtside) < 0)
	  goto no_match;
      }
      // success: save it
      success = it->from();
     no_match: ;
    }
  // if succeeded, save it
  if (success.live()) {
    _to_pp_from.push_back(houtside);
    _to_pp_to.push_back(success);
    return true;
  } else
    return false;
}

bool
Matcher::check_out_of(const PortT &hinside, const PortT &houtside)
{
  PortT phinside(_back_match[hinside.eindex()], hinside.port);
  PortT success;
  // now look for matches
  for (RouterT::conn_iterator it = _pat->find_connections_from(phinside);
       it != _pat->end_connections(); ++it)
    if (it->to_element() == _pat_output
	&& (success.dead() || it->to() < success)) {
      Vector<PortT> pto_pht, to_houtside;
      // check that it's valid: all connections to tunnels are covered
      // in body
      _pat->find_connections_to(it->to(), pto_pht);
      _body->find_connections_to(houtside, to_houtside);
      for (int j = 0; j < pto_pht.size(); j++) {
	PortT want(_match[pto_pht[j].eindex()], pto_pht[j].port);
	if (want.index_in(to_houtside) < 0)
	  goto no_match;
      }
      // success: save it
      success = it->to();
     no_match: ;
    }
  // if succeeded, save it
  if (success.live()) {
    _from_pp_from.push_back(success);
    _from_pp_to.push_back(houtside);
    return true;
  } else
    return false;
}

bool
Matcher::check_match()
{
  _from_pp_from.clear();
  _from_pp_to.clear();
  _to_pp_from.clear();
  _to_pp_to.clear();
  _defs.clear();

  // check configurations
  //fprintf(stderr, "CONF\n");
  for (int i = 0; i < _match.size(); i++)
    if (_match[i]) {
      if (!match_config(_pat->element(i)->configuration(), _match[i]->configuration(), _defs))
	return false;
    }

  int bnf = _body->nelements();
  _back_match.assign(bnf, 0);
  bool all_previous_match = true;
  for (int i = 0; i < _match.size(); i++)
    if (ElementT *m = _match[i]) {
      _back_match[m->eindex()] = _pat->element(i);
      if (m->flags != _patid)	// doesn't come from replacement of same pat
	all_previous_match = false;
    }

  // don't allow match if it consists entirely of elements previously replaced
  // by this pattern
  if (all_previous_match)
    return false;

  // find the pattern ports any cross-pattern jumps correspond to
  //fprintf(stderr, "XPJ\n");
  for (RouterT::conn_iterator it = _body->begin_connections();
       it != _body->end_connections(); ++it) {
    const PortT &hf = it->from(), &ht = it->to();
    ElementT *pf = _back_match[hf.eindex()], *pt = _back_match[ht.eindex()];
    if (pf && pt) {
      if (!_pat->has_connection(PortT(pf, hf.port), PortT(pt, ht.port)))
	return false;
    } else if (!pf && pt) {
      if (!check_into(hf, ht))
	return false;
    } else if (pf && !pt) {
      if (!check_out_of(hf, ht))
	return false;
    }
  }

  // check for unconnected tunnels in the pattern
  //fprintf(stderr, "UNC\n");
  for (RouterT::conn_iterator it = _pat->begin_connections();
       it != _body->end_connections(); ++it) {
    if (it->from_element() == _pat_input
	&& it->from().index_in(_to_pp_to) < 0)
      return false;
    if (it->to_element() == _pat_output
	&& it->to().index_in(_from_pp_from) < 0)
      return false;
  }

  //fprintf(stderr, "  YES\n");
  return true;
}

bool
Matcher::next_match()
{
  while (_pat_m->next_subgraph_isomorphism(_body_m, _match)) {
    //fprintf(stderr, "NEXT\n");
    if (check_subgraph_isomorphism(_pat, _body, _match)
	&& check_match())
      return true;
  }
  return false;
}



static String
uniqueify_prefix(const String &base_prefix, RouterT *r)
{
  // speed up uniqueification on the same prefix
  static HashTable<String, int> *last_uniqueifier;
  if (!last_uniqueifier)
    last_uniqueifier = new HashTable<String, int>(1);
  int count = last_uniqueifier->get(base_prefix);

  while (1) {
    String prefix = base_prefix + "@" + String(count);
    count++;

    // look for things starting with that name
    int plen = prefix.length();
    for (RouterT::iterator x = r->begin_elements(); x; x++) {
      const String &n = x->name();
      if (n.length() > plen + 1 && n.substring(0, plen) == prefix
	  && n[plen] == '/')
	goto failed;
    }

    last_uniqueifier->set(base_prefix, count);
    return prefix;

   failed: ;
  }
}

void
Matcher::replace_config(ElementT *e) const
{
  Vector<String> confvec;
  cp_argvec(e->configuration(), confvec);

  bool changed = false;
  for (int i = 0; i < confvec.size(); i++) {
    if (confvec[i].length() <= 1 || confvec[i][0] != '$')
      continue;
    if (HashTable<String, String>::const_iterator it = _defs.find(confvec[i])) {
      confvec[i] = it.value();
      changed = true;
    }
  }

  if (changed)
    e->set_configuration(cp_unargvec(confvec));
}

void
Matcher::replace(RouterT *replacement, const String &try_prefix,
		 const LandmarkT &landmark, ErrorHandler *errh)
{
  //fprintf(stderr, "replace...\n");
  String prefix = uniqueify_prefix(try_prefix, _body);

  // free old elements
  Vector<int> changed_elements;
  Vector<String> old_names;
  for (int i = 0; i < _match.size(); i++)
    if (_match[i]) {
      changed_elements.push_back(_match[i]->eindex());
      old_names.push_back(_match[i]->name());
      _body->free_element(_match[i]);
    } else
      old_names.push_back(String());

  // add replacement
  // collect new element indices in 'changed_elements'
  _body->set_new_eindex_collector(&changed_elements);

  // make element named 'prefix'
  ElementT *new_e = _body->get_element(prefix, ElementClassT::tunnel_type(), String(), landmark);

  // expand 'replacement' into '_body'; need crap compound element
  Vector<String> crap_args;
  VariableEnvironment crap_ve(0);
  replacement->complex_expand_element(new_e, crap_args, _body, String(), crap_ve, errh);

  // mark replacement
  for (int i = 0; i < changed_elements.size(); i++) {
    ElementT *e = _body->element(changed_elements[i]);
    if (e->dead())
      continue;
    e->flags = _patid;
    replace_config(e);
  }

  // save old element name if matched element and some replacement element
  // have the same name
  for (int i = 0; i < _match.size(); i++)
    if (_match[i]) {
      String n = _pat->ename(i);
      int new_index = _body->eindex(prefix + "/" + n);
      if (new_index >= 0)
	_body->change_ename(new_index, old_names[i]);
    }

  // find input and output, add connections to tunnels
  ElementT *new_pp = _body->element(prefix);
  for (int i = 0; i < _to_pp_from.size(); i++)
    _body->add_connection(_to_pp_from[i], PortT(new_pp, _to_pp_to[i].port),
			  landmark);
  for (int i = 0; i < _from_pp_from.size(); i++)
    _body->add_connection(PortT(new_pp, _from_pp_from[i].port),
			  _from_pp_to[i], landmark);

  // cleanup
  _body->remove_tunnels();
  // remember to clear 'new_eindex_collector'!
  _body->set_new_eindex_collector(0);
  _match.clear();

  // finally, update the adjacency matrix
  _body_m->update(changed_elements);
  //_body_m->init(_body);
}


// match configuration strings
bool
match_config(const String &pat, const String &conf,
	     HashTable<String, String> &defs)
{
  Vector<String> patvec, confvec;
  HashTable<String, String> my_defs;

  // separate into vectors
  cp_argvec(pat, patvec);
  cp_argvec(conf, confvec);

  // not same size -> no match
  if (patvec.size() != confvec.size())
    return false;

  // go over elements and compare; handle $variables
  for (int i = 0; i < patvec.size(); i++) {
    if (patvec[i] == confvec[i])
      continue;
    if (patvec[i].length() <= 1 || patvec[i][0] != '$')
      return false;
    const String &p = patvec[i];
    for (int j = 1; j < p.length(); j++)
      if (!isalnum((unsigned char) p[j]) && p[j] != '_')
	return false;
    if (HashTable<String, String>::iterator it = defs.find(p)) {
      if (it.value() != confvec[i])
	return false;
    } else if (HashTable<String, String>::iterator it = my_defs.find(p)) {
      if (it.value() != confvec[i])
	return false;
    } else
      my_defs.set(p, confvec[i]);
  }

  // insert my defs into defs
  for (HashTable<String, String>::iterator iter = my_defs.begin(); iter.live(); iter++)
    defs.set(iter.key(), iter.value());

  return true;
}


#define HELP_OPT		300
#define VERSION_OPT		301
#define ROUTER_OPT		303
#define EXPRESSION_OPT		304
#define OUTPUT_OPT		305
#define PATTERNS_OPT		306
#define REVERSE_OPT		307

static const Clp_Option options[] = {
  { "expression", 'e', EXPRESSION_OPT, Clp_ValString, 0 },
  { "file", 'f', ROUTER_OPT, Clp_ValString, 0 },
  { "help", 0, HELP_OPT, 0, 0 },
  { "output", 'o', OUTPUT_OPT, Clp_ValString, 0 },
  { "patterns", 'p', PATTERNS_OPT, Clp_ValString, 0 },
  { "reverse", 'r', REVERSE_OPT, 0, Clp_Negate },
  { "version", 'v', VERSION_OPT, 0, 0 },
};

static const char *program_name;

void
short_usage()
{
  fprintf(stderr, "Usage: %s [OPTION]... [ROUTERFILE] [PATTERNFILE]...\n\
Try '%s --help' for more information.\n",
	  program_name, program_name);
}

void
usage()
{
  printf("\
'Click-xform' replaces patterns of elements with other sets of elements inside\n\
a Click router configuration. Both patterns and configuration are Click-\n\
language files. The transformed router configuration is written to the\n\
standard output.\n\
\n\
Usage: %s [OPTION]... [ROUTERFILE] [PATTERNFILE]...\n\
\n\
Options:\n\
  -p, --patterns PATTERNFILE    Read patterns from PATTERNFILE. Can be given\n\
                                more than once.\n\
  -f, --file FILE               Read router configuration from FILE.\n\
  -e, --expression EXPR         Use EXPR as router configuration.\n\
  -o, --output FILE             Write output to FILE.\n\
  -r, --reverse                 Apply patterns in reverse.\n\
      --help                    Print this message and exit.\n\
  -v, --version                 Print version number and exit.\n\
\n\
Report bugs to <click@pdos.lcs.mit.edu>.\n", program_name);
}

static Vector<RouterT *> patterns;
static Vector<RouterT *> replacements;
static Vector<String> pat_names;
static int patterns_attempted;

void
read_pattern_file(const char *name, ErrorHandler *errh)
{
  patterns_attempted++;
  RouterT *pat_file = read_router_file(name, true, errh);
  if (!pat_file)
    return;

  Vector<ElementClassT *> compounds;
  pat_file->collect_locally_declared_types(compounds);

  for (int i = 0; i < compounds.size(); i++) {
    String name = compounds[i]->name();
    if (compounds[i]->cast_router() && name.length() > 12
	&& name.substring(-12) == "_Replacement") {
      ElementClassT *tt = pat_file->locally_declared_type(name.substring(0, -12));
      if (tt && tt->cast_router()) {
	RouterT *rep = compounds[i]->cast_router();
	RouterT *pat = tt->cast_router();
	if (rep && pat) {
	  patterns.push_back(pat);
	  replacements.push_back(rep);
	  pat_names.push_back(name.substring(0, -12));
	}
      }
    }
  }
}

int
main(int argc, char **argv)
{
  click_static_initialize();
  CLICK_DEFAULT_PROVIDES;
  ErrorHandler *errh = ErrorHandler::default_handler();

  // read command line arguments
  Clp_Parser *clp =
    Clp_NewParser(argc, argv, sizeof(options) / sizeof(options[0]), options);
  Clp_SetOptionChar(clp, '+', Clp_ShortNegated);
  program_name = Clp_ProgramName(clp);

  int num_nondash_args = 0;
  const char *router_file = 0;
  bool file_is_expr = false;
  const char *output_file = 0;
  bool reverse = 0;

  while (1) {
    int opt = Clp_Next(clp);
    switch (opt) {

     case HELP_OPT:
      usage();
      exit(0);
      break;

     case VERSION_OPT:
      printf("click-xform (Click) %s\n", CLICK_VERSION);
      printf("Copyright (c) 1999-2000 Massachusetts Institute of Technology\n\
This is free software; see the source for copying conditions.\n\
There is NO warranty, not even for merchantability or fitness for a\n\
particular purpose.\n");
      exit(0);
      break;

     case PATTERNS_OPT:
      read_pattern_file(clp->vstr, errh);
      break;

     case ROUTER_OPT:
     case EXPRESSION_OPT:
      if (router_file) {
	errh->error("router configuration specified twice");
	goto bad_option;
      }
      router_file = clp->vstr;
      file_is_expr = (opt == EXPRESSION_OPT);
      break;

     case OUTPUT_OPT:
      if (output_file) {
	errh->error("output file specified twice");
	goto bad_option;
      }
      output_file = clp->vstr;
      break;

     case REVERSE_OPT:
      reverse = !clp->negated;
      break;

     case Clp_NotOption:
      if (click_maybe_define(clp->vstr, errh))
	  break;
      if (num_nondash_args == 0 && router_file) {
	errh->error("router configuration specified twice");
	goto bad_option;
      } else if (num_nondash_args == 0)
	router_file = clp->vstr;
      else
	read_pattern_file(clp->vstr, errh);
      num_nondash_args++;
      break;

     bad_option:
     case Clp_BadOption:
      short_usage();
      exit(1);
      break;

     case Clp_Done:
      goto done;

    }
  }

 done:
  RouterT *r = read_router(router_file, file_is_expr, errh);
  if (r)
    r->flatten(errh);
  if (!r || errh->nerrors() > 0)
    exit(1);

  if (!patterns_attempted)
    errh->warning("no patterns read");

  // manipulate patterns
  if (patterns.size()) {
    // reverse if necessary
    if (reverse) {
      Vector<RouterT *> new_patterns, new_replacements;
      for (int i = patterns.size() - 1; i >= 0; i--) {
	new_patterns.push_back(replacements[i]);
	new_replacements.push_back(patterns[i]);
      }
      patterns.swap(new_patterns);
      replacements.swap(new_replacements);
    }

    // flatten patterns
    for (int i = 0; i < patterns.size(); i++)
      patterns[i]->flatten(errh);
  }

  // clear r's flags, so we know the current element complement
  // didn't come from replacements (paranoia)
  for (int i = 0; i < r->nelements(); i++)
    r->element(i)->flags = 0;

  // get adjacency matrices
  Vector<AdjacencyMatrix *> patterns_adj;
  for (int i = 0; i < patterns.size(); i++)
    patterns_adj.push_back(new AdjacencyMatrix(patterns[i]));

  bool any = true;
  int nreplace = 0;
  AdjacencyMatrix matrix(r);
  while (any) {
    any = false;
    for (int i = 0; i < patterns.size(); i++) {
      Matcher m(patterns[i], patterns_adj[i], r, &matrix, i + 1, errh);
      if (m.next_match()) {
	m.replace(replacements[i], pat_names[i], LandmarkT(), errh);
	nreplace++;
	any = true;
	break;
      }
    }
  }

  // write result
  if (nreplace)
    r->remove_dead_elements();
  if (write_router_file(r, output_file, errh) < 0)
    exit(1);
  return 0;
}
